base/set.jl

```diff
@@ -191,6 +191,62 @@ function unique(f::Callable, C)
 end
 
 """
+    unique!(c)
+Remove duplicate items and return the modified collection, as determined by
+[`isequal`](@ref).
+
+```jldoctest
+julia> unique!([1, 1, 1])
+1-element Array{Int64,1}:
+ 1
+
+julia> A = [7, 3, 2, 3, 7, 5];
+
+julia> unique!(A);
+
+julia> A
+4-element Array{Int64,1}:
+ 7
+ 3
+ 2
+ 5
+```
+"""
+function unique!(c)
+    if issorted(c) || issorted(reverse(c))
+        idxs = eachindex(c)
+        # When we loop through c, we only need to remember one state and one
+        # item in the collection.
+        currentstate = start(idxs)
+        currentitem = c[currentstate]
```

Comment:
Thanks for all of your help, from what I understand from above, this is what I managed to hack together
```julia
function unique!(c)
    if issorted(c) || issorted(c, rev=true)
        done(c, 1) ? count = 0 : count = 1
        for i in indices(c, 1)
            x = c[i]
            currentitem = c[count]
            if x != currentitem
                count += 1
                c[count] = x
            end
        end
    else
        seen = Set{eltype(c)}()
        idxs = eachindex(c)
        m = n = start(idxs)
        count = 0
        while !done(c, n)
            i, n = next(idxs, n)
            x = c[i]
            if x ∉ seen
                count += 1
                push!(seen, x)
                j, m = next(idxs, m)
                c[j] = x
            end
        end
    end
    resize!(c, count)
end
```